compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/AssertionLowering.kt

```diff
@@ -0,0 +1,156 @@
+            at(expression)
+            val assertCondition = expression.getValueArgument(0)!!
+            val lambdaArgument = if (function.valueParameters.size == 2) expression.getValueArgument(1) else null
+
+            return if (mode == JVMAssertionsMode.ALWAYS_ENABLE) {
+                checkAssertion(assertCondition, lambdaArgument)
+            } else {
+                require(mode == JVMAssertionsMode.JVM && data != null)
+                irIfThen(
+                    irNot(getAssertionDisabled(this, data)),
+                    checkAssertion(assertCondition, lambdaArgument)
+                )
+            }
+        }
+    }
+
+    private fun IrBuilderWithScope.checkAssertion(assertCondition: IrExpression, lambdaArgument: IrExpression?) =
+        irBlock {
+            val lambda = lambdaArgument?.asSimpleLambda()
+            val invokeVar = if (lambda == null && lambdaArgument != null) irTemporary(lambdaArgument) else null
+
+            val constructor = this@AssertionLowering.context.ir.symbols.assertionErrorConstructor
+            val throwError = irThrow(irCall(constructor).apply {
+                putValueArgument(
+                    0,
+                    when {
+                        lambda != null -> lambda.inline()
+                        lambdaArgument != null -> {
+                            val invoke =
+                                lambdaArgument.type.getClass()!!.functions.single { it.name == OperatorNameConventions.INVOKE }
+                            irCallOp(invoke.symbol, invoke.returnType, irGet(invokeVar!!))
+                        }
+                        else -> irString("Assertion failed")
+                    }
+                )
+            })
+            +irIfThen(irNot(assertCondition), throwError)
+        }
+
+    fun getAssertionDisabled(irBuilder: IrBuilderWithScope, data: ClassInfo): IrExpression {
+        if (data.assertionsDisabledField == null)
+            data.assertionsDisabledField = buildAssertionsDisabledField(data.irClass)
+        return irBuilder.irGetField(null, data.assertionsDisabledField!!)
+    }
+
+    private fun buildAssertionsDisabledField(irClass: IrClass) =
+        buildField {
+            name = Name.identifier("\$assertionsDisabled")
+            origin = JvmLoweredDeclarationOrigin.GENERATED_ASSERTION_ENABLED_FIELD
+            type = context.irBuiltIns.booleanType
+            isFinal = true
+            isStatic = true
+        }.apply {
+            parent = irClass
+            initializer = context.createIrBuilder(irClass.symbol).run {
+                at(this@apply)
+                irExprBody(irNot(irCall(this@AssertionLowering.context.ir.symbols.desiredAssertionStatus).apply {
+                    dispatchReceiver = getJavaClass(irClass)
+                }))
+            }
+        }
+
+    private fun getJavaClass(irClass: IrClass) =
+        context.buildJavaClassReference(context.state.typeMapper.mapClass(irClass.descriptor))
```

Comment:
Could you please avoid type mapping here and generate a proper class reference with the default type of the class, instead of CrIrType of its ASM type? I've removed CrIrType in fc5e9f5c60 (not in master yet) since it was pretty much a hack, and would be grateful if it wasn't used in more places

Reply:
Absolutely! This was actually on my todo list anyway. I'll fix it as soon as your commit is merged.

Reply:
Done!